// @serenity-js/playwright/esm/screenplay/models/PlaywrightBrowsingSessionWithBrowser.js
import { PlaywrightBrowsingSession } from './PlaywrightBrowsingSession.js';
/**
 *  @group Models
 */
export class PlaywrightBrowsingSessionWithBrowser extends PlaywrightBrowsingSession {
    browser;
    browserContextOptions;
    constructor(browser, browserContextOptions, extraBrowserContextOptions, selectors) {
        super(extraBrowserContextOptions, selectors);
        this.browser = browser;
        this.browserContextOptions = browserContextOptions;
    }
    async createBrowserContext() {
        // todo: move to initialise?
        return this.browser.newContext(this.browserContextOptions);
    }
    async registerCurrentPage() {
        const context = await this.browserContext();
        await context.newPage();
        // calling context.newPage() triggers a callback registered via browserContext(),
        // which wraps playwright.Page in PlaywrightPage and adds it to the list of pages
        // returned by this.allPages()
        const allPages = await this.allPages();
        return allPages.at(-1);
    }
    async browserCapabilities() {
        return {
            browserName: this.browser._initializer.name, // todo: raise a PR to Playwright to expose this information
            platformName: process.platform, // todo: get the actual platform from Playwright
            browserVersion: this.browser.version()
        };
    }
    async discard() {
        await this.closeAllPages();
        const context = await this.browserContext();
        await context.close();
    }
}
