// @serenity-js/core/esm/model/tags/Tags.js
import { ensure, isDefined, isString, match } from 'tiny-types';
import { ArbitraryTag } from './ArbitraryTag.js';
import { BrowserTag } from './BrowserTag.js';
import { CapabilityTag } from './CapabilityTag.js';
import { ExecutionRetriedTag } from './ExecutionRetriedTag.js';
import { FeatureTag } from './FeatureTag.js';
import { IssueTag } from './IssueTag.js';
import { ManualTag } from './ManualTag.js';
import { PlatformTag } from './PlatformTag.js';
import { ProjectTag } from './ProjectTag.js';
import { ThemeTag } from './ThemeTag.js';
function hasCustomDeserialiser(tagType) {
    return Object.prototype.hasOwnProperty.call(tagType, 'fromJSON');
}
/**
 * @package
 */
export class Tags {
    static supportedTypes = new Map([
        ArbitraryTag,
        BrowserTag,
        CapabilityTag,
        ExecutionRetriedTag,
        FeatureTag,
        IssueTag,
        ManualTag,
        PlatformTag,
        ProjectTag,
        ThemeTag,
    ].map(tagType => [tagType.Type, tagType]));
    static Pattern = /^@([\w-]+)[\s:]?(.*)/i;
    static matchTags(tagText) {
        if (tagText === '') {
            return [];
        }
        const [, tagType, value] = Tags.Pattern.exec(tagText);
        return match(tagType.toLowerCase())
            .when('manual', _ => [new ManualTag()])
            .when(/^issues?$/, _ => value.split(',').map(value => new IssueTag(value.trim())))
            .else(value => [new ArbitraryTag(value.trim())]);
    }
    static fromJSON(o) {
        const type = ensure('serialised tag type', o.type, isDefined(), isString());
        if (!this.supportedTypes.has(type)) {
            return new ArbitraryTag(o.name);
        }
        const found = this.supportedTypes.get(type);
        if (hasCustomDeserialiser(found)) {
            return found.fromJSON(o);
        }
        return new found(o.name);
    }
    static from(text) {
        const tags = text.split(/\s+/)
            .filter(word => word.startsWith('@'))
            .flatMap(tag => Tags.matchTags(tag));
        return tags ?? Tags.matchTags(text);
    }
    static stripFrom(text) {
        return text.split(/\s+/)
            .map(word => Tags.Pattern.test(word) ? undefined : word)
            .filter(Boolean)
            .join(' ')
            .trim();
    }
}
