// @serenity-js/cucumber/esm/listeners/messages/CucumberMessagesListener.js
import { IdGenerator } from '@cucumber/messages';
import { SceneFinishes, TestRunFinished, TestRunFinishes, TestRunStarts } from '@serenity-js/core/events';
import { ExecutionFailedWithError, ExecutionSuccessful } from '@serenity-js/core/model';
import { CucumberMessagesParser } from './parser/CucumberMessagesParser.js';
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export default function (serenity, moduleLoader) {
    const { Formatter, formatterHelpers } = moduleLoader.require('@cucumber/cucumber'), TestCaseHookDefinition = moduleLoader.require('@cucumber/cucumber/lib/models/test_case_hook_definition').default;
    return class CucumberMessagesListener extends Formatter {
        static fakeInternalAfterHookUri = '/internal/serenity-js/cucumber';
        parser;
        log;
        supportCodeLibrary;
        constructor(options) {
            super(options);
            this.parser = new CucumberMessagesParser(serenity, formatterHelpers, options, (step) => step?.actionLocation?.uri !== CucumberMessagesListener.fakeInternalAfterHookUri);
            this.supportCodeLibrary = this.supportCodeLibrary ?? options.supportCodeLibrary;
            this.addAfterHook(() => {
                this.emit(new SceneFinishes(serenity.currentSceneId(), new ExecutionSuccessful(), serenity.currentTime()));
                return serenity.waitForNextCue();
            });
            options.eventBroadcaster.on('envelope', (envelope) => {
                // this.log('> [cucumber] ' + JSON.stringify(envelope) + '\n');
                switch (true) {
                    case !!envelope.testRunStarted:
                        return this.emit(new TestRunStarts(serenity.currentTime()));
                    case !!envelope.testCaseStarted:
                        return this.emit(this.parser.parseTestCaseStarted(envelope.testCaseStarted));
                    case !!envelope.testStepStarted:
                        return this.emit(this.parser.parseTestStepStarted(envelope.testStepStarted));
                    case !!envelope.testStepFinished:
                        return this.emit(this.parser.parseTestStepFinished(envelope.testStepFinished));
                    case !!envelope.testCaseFinished:
                        return this.emit(this.parser.parseTestCaseFinished(envelope.testCaseFinished));
                }
            });
        }
        async finished() {
            this.emit(new TestRunFinishes(serenity.currentTime()));
            try {
                await serenity.waitForNextCue();
                this.emit(new TestRunFinished(new ExecutionSuccessful(), serenity.currentTime()));
            }
            catch (error) {
                this.emit(new TestRunFinished(new ExecutionFailedWithError(error), serenity.currentTime()));
                throw error;
            }
            finally {
                await super.finished();
            }
        }
        addAfterHook(code) {
            this.supportCodeLibrary.afterTestCaseHookDefinitions.unshift(new TestCaseHookDefinition({
                code,
                id: IdGenerator.uuid()(),
                line: 0,
                uri: CucumberMessagesListener.fakeInternalAfterHookUri,
                options: {},
            }));
        }
        emit(events) {
            [].concat(events).forEach(event => serenity.announce(event));
        }
    };
}
