// @serenity-js/core/lib/stage/queues/DomainEventQueues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DomainEventQueues = void 0;
const index_js_1 = require("../../errors/index.js");
const index_js_2 = require("../../events/index.js");
const format_js_1 = require("../../io/format.js");
const DomainEventQueue_js_1 = require("./DomainEventQueue.js");
/**
 * @group Stage
 */
class DomainEventQueues {
    queueIndex = [];
    queues = new Map();
    holdingBay = new DomainEventQueue_js_1.DomainEventQueue();
    enqueue(event) {
        if (this.shouldStartNewQueueFor(event)) {
            this.queues.set(this.queueIdFor(event), new DomainEventQueue_js_1.DomainEventQueue(event, ...this.holdingBay.drain()));
        }
        else if (this.hasNoQueuesReadyFor(event)) {
            this.holdingBay.enqueue(event);
        }
        else {
            this.queues.get(this.queueIdFor(event)).enqueue(event);
        }
    }
    forEach(callback) {
        this.queues.forEach(callback);
    }
    drainQueueFor(sceneId) {
        for (const [key, queue] of this.queues) {
            if (sceneId.equals(queue.sceneId)) {
                const events = queue.drain();
                this.queues.delete(key);
                return events;
            }
        }
        throw new index_js_1.LogicError((0, format_js_1.d) `No domain event queue found for scene ${sceneId}`);
    }
    shouldStartNewQueueFor(event) {
        return (event instanceof index_js_2.SceneSequenceDetected || event instanceof index_js_2.SceneStarts)
            && !this.queues.has(this.queueIdFor(event));
    }
    hasNoQueuesReadyFor(event) {
        return this.queues.size === 0
            || !this.queues.has(this.queueIdFor(event));
    }
    queueIdFor(event) {
        const exactMatch = this.queueIndex.find(entry => entry.sceneId.equals(event.sceneId));
        if (exactMatch) {
            return exactMatch.queueId;
        }
        if (!(event instanceof index_js_2.SceneStarts)) {
            const sameScenarioMatch = this.queueIndex.find(entry => entry.details &&
                entry.details.equals(event.details));
            if (sameScenarioMatch) {
                this.queueIndex.push({
                    sceneId: event.sceneId,
                    details: event.details || sameScenarioMatch.details,
                    queueId: sameScenarioMatch.queueId,
                });
                return sameScenarioMatch.queueId;
            }
        }
        const newQueueId = Symbol();
        this.queueIndex.push({
            sceneId: event.sceneId,
            details: event.details,
            queueId: newQueueId,
        });
        return newQueueId;
    }
}
exports.DomainEventQueues = DomainEventQueues;
