// @serenity-js/cucumber/esm/listeners/legacy/notifier/Notifier.js
import { FeatureNarrativeDetected, SceneDescriptionDetected, SceneFinished, SceneFinishes, SceneParametersDetected, SceneSequenceDetected, SceneStarts, SceneTagged, SceneTemplateDetected, TaskFinished, TaskStarts, TestRunFinished, TestRunFinishes, TestRunnerDetected, TestRunStarts, } from '@serenity-js/core/events';
import { ActivityDetails, CapabilityTag, Category, Description, ExecutionSuccessful, FeatureTag, Name, ScenarioDetails, Tag, ThemeTag } from '@serenity-js/core/model';
function notEmpty(list) {
    return list.filter(item => !!item);
}
/**
 * @private
 */
export class Notifier {
    serenity;
    currentSceneId;
    currentScenario;
    currentStepActivityId;
    constructor(serenity) {
        this.serenity = serenity;
    }
    testRunStarts() {
        this.emit(new TestRunStarts(this.serenity.currentTime()));
    }
    outlineDetected(sceneId, scenario, outline, feature) {
        const outlineDetails = this.detailsOf(outline, feature), scenarioDetails = this.detailsOf(scenario, feature), template = outline.steps.map(step => step.name.value).join('\n');
        this.emit(...notEmpty([
            new SceneSequenceDetected(sceneId, outlineDetails, this.serenity.currentTime()),
            new SceneTemplateDetected(sceneId, new Description(template), this.serenity.currentTime()),
            new SceneParametersDetected(sceneId, scenarioDetails, outline.parameters[scenario.location.line], this.serenity.currentTime()),
        ]));
    }
    scenarioStarts(sceneId, scenario, feature) {
        this.currentSceneId = sceneId;
        const details = this.detailsOf(scenario, feature);
        this.currentScenario = details;
        // todo: emit SceneBackgroundDetected?
        this.emit(...notEmpty([
            new SceneStarts(this.currentSceneId, details, this.serenity.currentTime()),
            feature.description && new FeatureNarrativeDetected(this.currentSceneId, feature.description, this.serenity.currentTime()),
            new TestRunnerDetected(this.currentSceneId, new Name('JS'), this.serenity.currentTime()),
            ...this.scenarioHierarchyTagsFor(feature).map(tag => new SceneTagged(this.currentSceneId, tag, this.serenity.currentTime())),
            !!scenario.description && new SceneDescriptionDetected(this.currentSceneId, scenario.description, this.serenity.currentTime()),
            ...scenario.tags.map(tag => new SceneTagged(this.currentSceneId, tag, this.serenity.currentTime())),
        ]));
    }
    stepStarts(step) {
        const activityDetails = new ActivityDetails(step.name, step.location);
        this.currentStepActivityId = this.serenity.assignNewActivityId(activityDetails);
        this.emit(new TaskStarts(this.currentSceneId, this.currentStepActivityId, activityDetails, this.serenity.currentTime()));
    }
    stepFinished(step, outcome) {
        this.emit(new TaskFinished(this.currentSceneId, this.currentStepActivityId, new ActivityDetails(step.name, step.location), outcome, this.serenity.currentTime()));
    }
    scenarioFinishes() {
        this.emitSceneFinishes();
    }
    scenarioFinished(scenario, feature, outcome) {
        const details = this.detailsOf(scenario, feature);
        this.emit(new SceneFinished(this.currentSceneId, details, outcome, this.serenity.currentTime()));
    }
    testRunFinishes() {
        this.emit(new TestRunFinishes(this.serenity.currentTime()));
    }
    testRunFinished(outcome) {
        this.emit(new TestRunFinished(outcome, this.serenity.currentTime()));
    }
    emitSceneFinishes() {
        this.emit(new SceneFinishes(this.currentSceneId, new ExecutionSuccessful(), this.serenity.currentTime()));
    }
    detailsOf(scenario, feature) {
        return new ScenarioDetails(scenario.name, new Category(feature.name.value), scenario.location);
    }
    scenarioHierarchyTagsFor(feature) {
        const directories = notEmpty(feature.location.path.directory().split()), featuresIndex = directories.indexOf('features'), hierarchy = [...directories.slice(featuresIndex + 1), feature.name.value];
        const [featureName, capabilityName, themeName] = hierarchy.reverse();
        return notEmpty([
            themeName && Tag.humanReadable(ThemeTag, themeName),
            capabilityName && Tag.humanReadable(CapabilityTag, capabilityName),
            feature && new FeatureTag(featureName),
        ]);
    }
    emit(...events) {
        events.forEach(event => this.serenity.announce(event));
    }
}
