// @serenity-js/cucumber/esm/listeners/legacy/notifier/ResultMapper.js
import { AssertionError, ImplementationPendingError, TestCompromisedError } from '@serenity-js/core';
import { ExecutionCompromised, ExecutionFailedWithAssertionError, ExecutionFailedWithError, ExecutionSkipped, ExecutionSuccessful, ImplementationPending } from '@serenity-js/core/model';
import { AmbiguousStepDefinitionError } from '../../../errors/index.js';
/**
 * @package
 */
export class ResultMapper {
    serenity;
    constructor(serenity) {
        this.serenity = serenity;
    }
    outcomeFor(status, maybeError) {
        const error = this.errorFrom(maybeError);
        if (this.isTimeoutError(error)) {
            return new ExecutionFailedWithError(error);
        }
        switch (true) {
            case status === 'undefined':
                return new ImplementationPending(new ImplementationPendingError('Step not implemented'));
            case status === 'ambiguous':
                if (!error) {
                    // Only the step result contains the "ambiguous step def error", the scenario itself doesn't
                    return new ExecutionFailedWithError(new AmbiguousStepDefinitionError('Multiple step definitions match'));
                }
                return new ExecutionFailedWithError(error);
            case status === 'failed':
                return this.problemIndicationOutcomeFromError(error);
            case status === 'pending':
                return new ImplementationPending(new ImplementationPendingError('Step not implemented'));
            case status === 'skipped':
                return new ExecutionSkipped();
            // case status === 'passed':
            default:
                return new ExecutionSuccessful();
        }
    }
    errorFrom(error) {
        switch (typeof error) {
            case 'string': return new Error(error);
            case 'object': return error;
            case 'function': return error;
            default: return void 0;
        }
    }
    isTimeoutError(error) {
        return error && /timed out/.test(error.message);
    }
    isANonSerenityAssertionError(error) {
        return error instanceof Error
            && error.name === 'AssertionError'
            && error.message && hasOwnProperty(error, 'expected')
            && hasOwnProperty(error, 'actual');
    }
    problemIndicationOutcomeFromError(error) {
        if (error instanceof AssertionError) {
            return new ExecutionFailedWithAssertionError(error);
        }
        if (this.isANonSerenityAssertionError(error)) {
            return new ExecutionFailedWithAssertionError(this.serenity.createError(AssertionError, {
                message: error.message,
                diff: {
                    expected: error.expected,
                    actual: error.actual,
                },
                cause: error,
            }));
        }
        if (error instanceof TestCompromisedError) {
            return new ExecutionCompromised(error);
        }
        return new ExecutionFailedWithError(error);
    }
}
/**
 * @private
 */
function hasOwnProperty(value, fieldName) {
    return Object.prototype.hasOwnProperty.call(value, fieldName);
}
