// @serenity-js/core/lib/io/loader/ModuleLoader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleLoader = void 0;
const node_module_1 = require("node:module");
const node_path_1 = __importDefault(require("node:path"));
const node_url_1 = require("node:url");
const Version_js_1 = require("../Version.js");
// Get the Module object for internal resolution APIs
const bootstrapRequire = (0, node_module_1.createRequire)((0, node_url_1.pathToFileURL)(node_path_1.default.join(process.cwd(), 'package.json')).href);
const Module = bootstrapRequire('module'); // No type definitions available
/**
 * Creates a require function for the given directory.
 * Uses createRequire to create a require function anchored to the specified cwd.
 */
function createRequireForCwd(cwd) {
    const contextFile = node_path_1.default.join(cwd, 'noop.js');
    return (0, node_module_1.createRequire)((0, node_url_1.pathToFileURL)(contextFile).href);
}
/**
 * Dynamically loads Node modules located relative to `cwd`.
 */
class ModuleLoader {
    cwd;
    useRequireCache;
    requireFn;
    /**
     * @param {string} cwd
     *  Current working directory, relative to which Node modules should be resolved.
     * @param useRequireCache
     *  Whether to use Node's `require.cache`. Defaults to `true`.
     *  Set to `false` to force Node to reload required modules on every call.
     */
    constructor(cwd, useRequireCache = true) {
        this.cwd = cwd;
        this.useRequireCache = useRequireCache;
        // Create a require function anchored to the cwd
        this.requireFn = createRequireForCwd(cwd);
    }
    /**
     * Returns `true` if a given module is available to be required, false otherwise.
     *
     * @param moduleId
     *  NPM module id, for example 'cucumber' or '@serenity-js/core'
     */
    hasAvailable(moduleId) {
        try {
            return !!this.require(moduleId);
        }
        catch {
            return false;
        }
    }
    /**
     * Works like `require.resolve`, but relative to specified current working directory
     *
     * @param moduleId
     *  NPM module id, for example `cucumber` or `@serenity-js/core`
     *
     * @returns
     *  Path to a given Node.js module
     */
    resolve(moduleId) {
        const fromFile = node_path_1.default.join(this.cwd, 'noop.js');
        return Module._resolveFilename(moduleId, {
            id: fromFile,
            filename: fromFile,
            paths: Module._nodeModulePaths(this.cwd),
        });
    }
    /**
     * Works like `require`, but relative to specified current working directory
     *
     * @param moduleId
     */
    require(moduleId) {
        try {
            return this.requireFn(this.cachedIfNeeded(this.resolve(moduleId)));
        }
        catch {
            return this.requireFn(this.cachedIfNeeded(moduleId));
        }
    }
    cachedIfNeeded(moduleId) {
        if (!this.useRequireCache) {
            delete this.requireFn.cache[moduleId];
        }
        return moduleId;
    }
    /**
     * Returns `Version` of module specified by `moduleId`, based on its `package.json`.
     *
     * @param moduleId
     */
    versionOf(moduleId) {
        return new Version_js_1.Version(this.require(`${moduleId}/package.json`).version);
    }
}
exports.ModuleLoader = ModuleLoader;
