// @serenity-js/cucumber/lib/listeners/legacy/gherkin/GherkinParserAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GherkinParserAdapter = void 0;
const gherkin_1 = require("@cucumber/gherkin");
const messages_1 = require("@cucumber/messages");
/**
 * Adapts the new @cucumber/gherkin Parser to produce output compatible with the old gherkin 5.x format.
 *
 * @private
 */
class GherkinParserAdapter {
    parser;
    constructor() {
        const createUuid = messages_1.IdGenerator.uuid();
        const builder = new gherkin_1.AstBuilder(createUuid);
        const matcher = new gherkin_1.GherkinClassicTokenMatcher();
        this.parser = new gherkin_1.Parser(builder, matcher);
    }
    parse(featureContent) {
        const newDoc = this.parser.parse(featureContent);
        return this.adaptDocument(newDoc);
    }
    adaptDocument(doc) {
        return {
            type: 'GherkinDocument',
            feature: doc.feature ? this.adaptFeature(doc.feature) : undefined,
            comments: (doc.comments || []).map(c => this.adaptComment(c)),
        };
    }
    adaptFeature(feature) {
        return {
            type: 'Feature',
            location: feature.location,
            tags: feature.tags.map(t => this.adaptTag(t)),
            language: feature.language,
            keyword: feature.keyword,
            name: feature.name,
            description: this.trimDescription(feature.description),
            children: feature.children.map(child => this.adaptChild(child)),
        };
    }
    adaptChild(child) {
        if (child.background) {
            return this.adaptBackground(child.background);
        }
        if (child.scenario) {
            // Check if it's a Scenario Outline by looking at the keyword or examples
            if (child.scenario.keyword === 'Scenario Outline' || child.scenario.keyword === 'Scenario Template' || child.scenario.examples.length > 0) {
                return this.adaptScenarioOutline(child.scenario);
            }
            return this.adaptScenario(child.scenario);
        }
        // Rule children are not supported in the legacy format
        throw new Error('Unsupported feature child type');
    }
    adaptBackground(bg) {
        return {
            type: 'Background',
            location: bg.location,
            tags: [],
            keyword: bg.keyword,
            name: bg.name,
            description: this.trimDescription(bg.description),
            steps: bg.steps.map(s => this.adaptStep(s)),
        };
    }
    adaptScenario(scenario) {
        return {
            type: 'Scenario',
            location: scenario.location,
            tags: scenario.tags.map(t => this.adaptTag(t)),
            keyword: scenario.keyword,
            name: scenario.name,
            description: this.trimDescription(scenario.description),
            steps: scenario.steps.map(s => this.adaptStep(s)),
        };
    }
    adaptScenarioOutline(scenario) {
        return {
            type: 'ScenarioOutline',
            location: scenario.location,
            tags: scenario.tags.map(t => this.adaptTag(t)),
            keyword: scenario.keyword,
            name: scenario.name,
            description: this.trimDescription(scenario.description),
            steps: scenario.steps.map(s => this.adaptStep(s)),
            examples: scenario.examples.map(e => this.adaptExamples(e)),
        };
    }
    adaptStep(step) {
        const result = {
            type: 'Step',
            location: step.location,
            keyword: step.keyword,
            text: step.text,
        };
        if (step.dataTable) {
            result.argument = this.adaptDataTable(step.dataTable);
        }
        else if (step.docString) {
            result.argument = this.adaptDocString(step.docString);
        }
        return result;
    }
    adaptDataTable(dataTable) {
        return {
            type: 'DataTable',
            location: dataTable.location,
            rows: dataTable.rows.map(r => this.adaptTableRow(r)),
        };
    }
    adaptDocString(docString) {
        return {
            type: 'DocString',
            location: docString.location,
            content: docString.content,
        };
    }
    adaptExamples(examples) {
        return {
            type: 'Examples',
            location: examples.location,
            tags: examples.tags.map(t => this.adaptTag(t)),
            keyword: examples.keyword,
            name: examples.name,
            description: this.trimDescription(examples.description),
            tableHeader: this.adaptTableRow(examples.tableHeader),
            tableBody: examples.tableBody.map(r => this.adaptTableRow(r)),
        };
    }
    adaptTableRow(row) {
        return {
            type: 'TableRow',
            location: row.location,
            cells: row.cells.map(c => ({
                type: 'TableCell',
                location: c.location,
                value: c.value,
            })),
        };
    }
    adaptTag(tag) {
        return {
            type: 'Tag',
            location: tag.location,
            name: tag.name,
        };
    }
    adaptComment(comment) {
        return {
            type: 'Comment',
            location: comment.location,
            text: comment.text,
        };
    }
    /**
     * Trims the description and returns it, or undefined if the description is empty.
     * The old gherkin 5.x parser returned undefined for empty descriptions, which the mapper
     * then passes directly to constructors using `description && new Description(description)`.
     */
    trimDescription(description) {
        const trimmed = description?.trim();
        // Return undefined for empty descriptions to match old gherkin 5.x behavior
        return trimmed || undefined;
    }
}
exports.GherkinParserAdapter = GherkinParserAdapter;
