// @serenity-js/web/esm/screenplay/models/Cookie.js
import { Interaction, Question, the, Timestamp } from '@serenity-js/core';
import { ensure, isBoolean, isDefined, isInstanceOf, isOneOf, isPlainObject, isString } from 'tiny-types';
import { CookieMissingError } from '../../errors/index.js';
import { BrowseTheWeb } from '../abilities/index.js';
/**
 * A Screenplay Pattern-style model responsible for managing cookies available to the current [`Page`](https://serenity-js.org/api/web/class/Page/).
 *
 * ## Checking if a cookie exists
 *
 * ```typescript
 * import { actorCalled } from '@serenity-js/core'
 * import { Navigate, Cookie } from '@serenity-js/web'
 * import { Ensure, isPresent } from '@serenity-js/assertions'
 *
 * await actorCalled('Sid')
 *   .attemptsTo(
 *     Navigate.to('https://example.org'),
 *
 *     Ensure.that(
 *       Cookie.called('example-cookie-name'),
 *       isPresent()
 *     ),
 * )
 * ```
 *
 * ## Setting a cookie
 *
 * ```typescript
 * import { actorCalled } from '@serenity-js/core'
 * import { Navigate, Cookie } from '@serenity-js/web'
 * import { Ensure, isPresent, not } from '@serenity-js/assertions'
 *
 * await actorCalled('Sid')
 *   .attemptsTo(
 *     Navigate.to('https://example.org'),
 *
 *     Ensure.that(Cookie.called('example-cookie-name'), not(isPresent())),
 *
 *     Cookie.set({
 *       name:  'favourite',
 *       value: 'triple chocolate',
 *     }),
 *
 *     Ensure.that(Cookie.called('example-cookie-name'), isPresent()),
 *   )
 * ```
 *
 * ## Reading a cookie
 *
 * ```typescript
 * import { actorCalled } from '@serenity-js/core'
 * import { Navigate, Cookie } from '@serenity-js/web'
 * import { Ensure, equals } from '@serenity-js/assertions'
 *
 * await actorCalled('Sid')
 *   .attemptsTo(
 *     Navigate.to('https://example.org'),
 *
 *     Ensure.that(
 *       Cookie.called('some-cookie-name').value(),
 *       equals('triple chocolate')
 *     ),
 * )
 * ```
 *
 * ## Learn more
 * - [`CookieData`](https://serenity-js.org/api/web/interface/CookieData/)
 * - [`Page.cookie`](https://serenity-js.org/api/web/class/Page/#cookie)
 *
 * @group Models
 */
export class Cookie {
    cookieName;
    /**
     * Creates a [`QuestionAdapter`](https://serenity-js.org/api/core/#QuestionAdapter) that resolves to [`Cookie`](https://serenity-js.org/api/web/class/Cookie/) identified by `name`.
     *
     * @param name
     */
    static called(name) {
        return Question.about(`"${name}" cookie`, async (actor) => {
            const cookieName = await actor.answer(name);
            const page = await BrowseTheWeb.as(actor).currentPage();
            return page.cookie(cookieName);
        });
    }
    /**
     * Sets a cookie for the current [`Page`](https://serenity-js.org/api/web/class/Page/). Note that [`CookieData`](https://serenity-js.org/api/web/interface/CookieData/) can be either a plain-old JavaScript object, or an [`Answerable`](https://serenity-js.org/api/core/#Answerable) [`WithAnswerableProperties`](https://serenity-js.org/api/core/#WithAnswerableProperties).
     *
     * :::info
     * Make sure that the actor performing this interaction is on the page that should receive the cookie.
     * Because of browser security restrictions, an actor can't set a cookie for an arbitrary page without being on that page.
     * :::
     *
     * @param cookieData
     */
    static set(cookieData) {
        return Interaction.where(the `#actor sets cookie: ${cookieData}`, async (actor) => {
            const cookie = ensure('cookieData', await actor.answer(Question.fromObject(cookieData)), isDefined(), isPlainObject());
            const page = await BrowseTheWeb.as(actor).currentPage();
            const sanitisedCookieData = {
                name: ensure(`Cookie.set(cookieData.name)`, cookie.name, isDefined(), isString()),
                value: ensure(`Cookie.set(cookieData.value)`, cookie.value, isDefined(), isString()),
                path: ensureIfPresent(cookie, 'path', isString()),
                domain: ensureIfPresent(cookie, 'domain', isString()),
                secure: ensureIfPresent(cookie, 'secure', isBoolean()),
                httpOnly: ensureIfPresent(cookie, 'httpOnly', isBoolean()),
                expiry: ensureIfPresent(cookie, 'expiry', isInstanceOf(Timestamp)),
                sameSite: ensureIfPresent(cookie, 'sameSite', isOneOf('Lax', 'Strict', 'None')),
            };
            return page.setCookie(sanitisedCookieData);
        });
    }
    /**
     * Creates an [interaction](https://serenity-js.org/api/core/class/Interaction/) to delete all cookies available to the current [`Page`](https://serenity-js.org/api/web/class/Page/)..
     */
    static deleteAll() {
        return Interaction.where(`#actor deletes all cookies`, async (actor) => {
            const page = await BrowseTheWeb.as(actor).currentPage();
            await page.deleteAllCookies();
        });
    }
    cookie;
    constructor(cookieName) {
        this.cookieName = cookieName;
        ensure('browser', cookieName, isDefined());
    }
    /**
     * Returns the name of this cookie.
     */
    name() {
        return this.cookieName;
    }
    /**
     * Checks if a given cookie is set.
     *
     * #### Learn more
     * - [`Optional`](https://serenity-js.org/api/core/interface/Optional/)
     */
    async isPresent() {
        try {
            const cookie = await this.lazyLoadCookie();
            return cookie && cookie.name === this.cookieName;
        }
        catch (error) {
            if (error instanceof CookieMissingError) {
                return false;
            }
            throw error;
        }
    }
    /**
     * Returns the value of a given cookie.
     */
    async value() {
        const cookie = await this.lazyLoadCookie();
        return cookie.value;
    }
    /**
     * Returns the path of a given cookie, if any was set.
     */
    async path() {
        const cookie = await this.lazyLoadCookie();
        return cookie.path;
    }
    /**
     * Returns the domain of a given cookie, if any was set.
     */
    async domain() {
        const cookie = await this.lazyLoadCookie();
        return cookie.domain;
    }
    /**
     * Checks if a given cookie is HTTP-only.
     *
     * #### Learn more
     * - [Mozilla Developer Network: Restricting access to cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#restrict_access_to_cookies)
     */
    async isHttpOnly() {
        const cookie = await this.lazyLoadCookie();
        return cookie.httpOnly;
    }
    /**
     * Checks if a given cookie is secure.
     *
     * #### Learn more
     * - [Mozilla Developer Network: Restricting access to cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#restrict_access_to_cookies)
     */
    async isSecure() {
        const cookie = await this.lazyLoadCookie();
        return cookie.secure;
    }
    /**
     * Returns the expiry date of a given cookie
     *
     * #### Learn more
     * - [`Timestamp`](https://serenity-js.org/api/core/class/Timestamp/)
     */
    async expiry() {
        const cookie = await this.lazyLoadCookie();
        return cookie.expiry;
    }
    /**
     * Invokes `Cookie.read` and caches the result in memory.
     */
    async lazyLoadCookie() {
        if (!this.cookie) {
            this.cookie = await this.read();
        }
        return this.cookie;
    }
}
/**
 * @ignore
 * @private
 *
 * @param data
 * @param property
 * @param predicates
 */
function ensureIfPresent(data, property, ...predicates) {
    return data[property]
        ? ensure(`Cookie.set(cookieData.${property})`, data[property], ...predicates)
        : undefined;
}
