// @serenity-js/web/esm/screenplay/models/Locator.js
import util from 'node:util';
import { LogicError } from '@serenity-js/core';
import { inspectedObject } from '@serenity-js/core/io';
import { RootLocator } from './RootLocator.js';
import { ByCss, ById, ByTagName } from './selectors/index.js';
/**
 * [`Locator`](https://serenity-js.org/api/web/class/Locator/) uses a [`Selector`](https://serenity-js.org/api/web/class/Selector/) to locate a [`PageElement`](https://serenity-js.org/api/web/class/PageElement/) or [`PageElement`](https://serenity-js.org/api/web/class/PageElements/)
 * within the [`Page`](https://serenity-js.org/api/web/class/Page/).
 *
 * ## Learn more
 * - [`RootLocator`](https://serenity-js.org/api/web/class/RootLocator/)
 * - [`Page.locate`](https://serenity-js.org/api/web/class/Page/#locate)
 * - [`PageElement`](https://serenity-js.org/api/web/class/PageElement/)
 * - [`PageElement`](https://serenity-js.org/api/web/class/PageElements/)
 *
 * @group Models
 */
export class Locator extends RootLocator {
    parent;
    selector;
    constructor(parent, selector) {
        super();
        this.parent = parent;
        this.selector = selector;
    }
    async switchToFrame(element) {
        await this.parent.switchToFrame(element);
    }
    async switchToParentFrame() {
        await this.parent.switchToParentFrame();
    }
    async switchToMainFrame() {
        await this.parent.switchToMainFrame();
    }
    /**
     * Expresses [`ByCss`](https://serenity-js.org/api/web/class/ByCss/), [`ById`](https://serenity-js.org/api/web/class/ById/), or [`ByTagName`](https://serenity-js.org/api/web/class/ByTagName/) as a [`ByCss`](https://serenity-js.org/api/web/class/ByCss/) selector.
     *
     * @throws LogicError
     *  if the `selector` can't be expressed as [`ByCss`](https://serenity-js.org/api/web/class/ByCss/)
     *
     * @param selector
     * @protected
     */
    asCssSelector(selector) {
        if (selector instanceof ByCss) {
            return selector;
        }
        if (selector instanceof ById) {
            return new ByCss(`#${selector.value}`);
        }
        if (selector instanceof ByTagName) {
            return new ByCss(`${selector.value}`);
        }
        throw new LogicError(`${selector} can't be expressed as a CSS selector`);
    }
    toString() {
        return this.selector.toString();
    }
    /**
     * Custom [Node.js inspection method](https://nodejs.org/api/util.html#utilinspectcustom).
     */
    [util.inspect.custom] = inspectedObject(this, ['parent', 'selector']);
}
