// @serenity-js/web/esm/screenplay/models/PageElement.js
import { Question, the } from '@serenity-js/core';
import { ensure, isDefined } from 'tiny-types';
import { BrowseTheWeb } from '../abilities/index.js';
/**
 * Uses the [actor's](https://serenity-js.org/api/core/class/Actor/) [ability](https://serenity-js.org/api/core/class/Ability/) to [`BrowseTheWeb`](https://serenity-js.org/api/web/class/BrowseTheWeb/) to identify
 * a single Web element located by [`Selector`](https://serenity-js.org/api/web/class/Selector/).
 *
 * ## Learn more
 * - [Page Element Query Language](https://serenity-js.org/handbook/web-testing/page-element-query-language)
 * - [`Optional`](https://serenity-js.org/api/core/interface/Optional/)
 * - [`Switchable`](https://serenity-js.org/api/web/interface/Switchable/)
 *
 * @group Models
 */
export class PageElement {
    locator;
    static from(nativeElement) {
        return Question.about(`native page element`, async (actor) => {
            const currentPage = await BrowseTheWeb.as(actor).currentPage();
            return currentPage.createPageElement(nativeElement);
        });
    }
    static located(selector) {
        return Question.about(the `page element located ${selector}`, async (actor) => {
            const bySelector = await actor.answer(selector);
            const currentPage = await BrowseTheWeb.as(actor).currentPage();
            return currentPage.locate(bySelector);
        });
    }
    static of(childElement, parentElement) {
        return Question.about(the `${childElement} of ${parentElement}`, async (actor) => {
            const parent = await actor.answer(parentElement);
            const child = childElement.of(parent);
            return actor.answer(child);
        });
    }
    /**
     * A static method producing a [`MetaQuestion`](https://serenity-js.org/api/core/interface/MetaQuestion/) that can be used with [`PageElements.eachMappedTo`](https://serenity-js.org/api/web/class/PageElements/#eachMappedTo) method
     * to extract the HTML of each element in a collection.
     *
     * #### Example
     *
     * ```typescript
     * import { actorCalled, Log } from '@serenity-js/core'
     * import { Navigate, PageElement, By, Text } from '@serenity-js/web'
     * import { includes } from '@serenity-js/assertions'
     *
     * await actorCalled('Debbie').attemptsTo(
     *   Navigate.to('https://serenity-js.org'),
     *
     *   Log.the(
     *     PageElements.located(By.css('a'))
     *       .where(Text, includes('modular'))
     *       .eachMappedTo(PageElement.html())
     *   ),
     * )
     * ```
     */
    static html() {
        return {
            of: (pageElement) => Question.about(`outer HTML of ${pageElement}`, async (actor) => {
                const element = await actor.answer(pageElement);
                return element.html();
            })
        };
    }
    constructor(locator) {
        this.locator = locator;
        ensure('native element locator', locator, isDefined());
    }
    /**
     * An "escape hatch" providing access to the integration tool-specific implementation of a Web element.
     */
    async nativeElement() {
        return this.locator.nativeElement();
    }
    toString() {
        return `PageElement located ${this.locator.toString()}`;
    }
    /**
     * Returns a [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that resolves to `true` when the element
     * is present in the [Document Object Model (DOM)](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model),
     * `false` otherwise.
     */
    async isPresent() {
        return this.locator.isPresent();
    }
}
