// tiny-types/lib/objects/equal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.equal = equal;
const TinyType_js_1 = require("../TinyType.js");
const significantFields_js_1 = require("./significantFields.js");
/**
 * @access private
 */
function equal(v1, v2) {
    switch (true) {
        case !sameType(v1, v2):
            return false;
        case both(arePrimitives, v1, v2):
            return checkIdentityOf(v1, v2);
        case both(areObjects, v1, v2) && sameClass(v1, v2) && both(areDates, v1, v2):
            return checkTimestamps(v1, v2);
        case both(areObjects, v1, v2) && sameClass(v1, v2):
            return checkSignificantFieldsOf(v1, v2);
    }
    return false;
}
const areObjects = (_) => new Object(_) === _;
const areDates = (_) => _ instanceof Date;
const arePrimitives = (_) => !areObjects(_); // arrays are objects
function both(condition, v1, v2) {
    return condition(v1) && condition(v2);
}
const sameType = (v1, v2) => typeof v1 === typeof v2;
/**
 * Gets all class names in the prototype chain of an object.
 */
function getClassNamesInChain(instance) {
    const names = [];
    let proto = Object.getPrototypeOf(instance);
    while (proto !== null && proto.constructor) {
        names.push(proto.constructor.name);
        proto = Object.getPrototypeOf(proto);
    }
    return names;
}
/**
 * Checks if two objects are of the same class or related by inheritance.
 * For TinyType instances, checks the prototype chain by class name to handle ESM/CJS dual-package hazard.
 */
const sameClass = (v1, v2) => {
    // For TinyType instances, check if they share a class relationship
    if ((0, TinyType_js_1.isTinyType)(v1) && (0, TinyType_js_1.isTinyType)(v2)) {
        const v1Name = v1.constructor.name;
        const v2Name = v2.constructor.name;
        // Same class
        if (v1Name === v2Name) {
            return true;
        }
        // Check inheritance: v1's class is in v2's chain or vice versa
        const v1Chain = getClassNamesInChain(v1);
        const v2Chain = getClassNamesInChain(v2);
        return v2Chain.includes(v1Name) || v1Chain.includes(v2Name);
    }
    // Fall back to instanceof for non-TinyType objects
    return (v1.constructor && v2 instanceof v1.constructor) || (v2.constructor && v1 instanceof v2.constructor);
};
const sameLength = (v1, v2) => v1.length === v2.length;
function checkIdentityOf(v1, v2) {
    return v1 === v2;
}
function checkTimestamps(v1, v2) {
    return v1.getTime() === v2.getTime();
}
function checkSignificantFieldsOf(o1, o2) {
    const fieldsOfObject1 = (0, significantFields_js_1.significantFieldsOf)(o1), fieldsOfObject2 = (0, significantFields_js_1.significantFieldsOf)(o2);
    if (!sameLength(fieldsOfObject1, fieldsOfObject2)) {
        return false;
    }
    return fieldsOfObject1.reduce((previousFieldsAreEqual, field) => {
        const currentFieldIsEqual = equal(o1[field], o2[field]);
        return previousFieldsAreEqual && currentFieldIsEqual;
    }, true);
}
