// @serenity-js/cucumber/esm/listeners/messages/parser/CucumberMessagesParser.js
import { TestStepResultStatus } from '@cucumber/messages';
import { AssertionError, ErrorSerialiser, ImplementationPendingError, TestCompromisedError } from '@serenity-js/core';
import { BusinessRuleDetected, FeatureNarrativeDetected, RetryableSceneDetected, SceneDescriptionDetected, SceneFinished, SceneParametersDetected, SceneSequenceDetected, SceneStarts, SceneTagged, SceneTemplateDetected, TaskFinished, TaskStarts, TestRunnerDetected, } from '@serenity-js/core/events';
import { FileSystem, FileSystemLocation, Path } from '@serenity-js/core/io';
import { RequirementsHierarchy } from '@serenity-js/core/io';
import { ActivityDetails, ArbitraryTag, BusinessRule, Category, Description, ExecutionCompromised, ExecutionFailedWithAssertionError, ExecutionFailedWithError, ExecutionRetriedTag, ExecutionSkipped, ExecutionSuccessful, ImplementationPending, Name, ScenarioDetails, ScenarioParameters, Tags, } from '@serenity-js/core/model';
import { TestStepFormatter } from './TestStepFormatter.js';
/**
 * @package
 */
export class CucumberMessagesParser {
    serenity;
    formatterHelpers;
    shouldReportStep;
    testStepFormatter = new TestStepFormatter();
    currentScenario;
    currentStepActivityId;
    cwd;
    eventDataCollector;
    snippetBuilder;
    supportCodeLibrary;
    requirementsHierarchy;
    constructor(serenity, formatterHelpers, // eslint-disable-line @typescript-eslint/explicit-module-boundary-types
    formatterOptionsAndDependencies, shouldReportStep) {
        this.serenity = serenity;
        this.formatterHelpers = formatterHelpers;
        this.shouldReportStep = shouldReportStep;
        this.cwd = formatterOptionsAndDependencies.cwd;
        this.eventDataCollector = formatterOptionsAndDependencies.eventDataCollector;
        this.snippetBuilder = formatterOptionsAndDependencies.snippetBuilder;
        this.supportCodeLibrary = formatterOptionsAndDependencies.supportCodeLibrary;
        this.requirementsHierarchy = new RequirementsHierarchy(new FileSystem(Path.from(formatterOptionsAndDependencies.cwd)), formatterOptionsAndDependencies.parsedArgvOptions?.specDirectory && Path.from(formatterOptionsAndDependencies.parsedArgvOptions?.specDirectory));
    }
    parseTestCaseStarted(message) {
        const testCaseAttempt = this.eventDataCollector.getTestCaseAttempt(message.id), currentSceneId = this.serenity.assignNewSceneId();
        this.currentScenario = this.scenarioDetailsFor(testCaseAttempt.gherkinDocument, testCaseAttempt.pickle, this.formatterHelpers.PickleParser.getPickleLocation(testCaseAttempt));
        return [
            ...this.extract(this.outlineFrom(testCaseAttempt), (outline) => [
                new SceneSequenceDetected(currentSceneId, outline.details, this.serenity.currentTime()),
                new SceneTemplateDetected(currentSceneId, outline.template, this.serenity.currentTime()),
                new SceneParametersDetected(currentSceneId, this.currentScenario, outline.parameters, this.serenity.currentTime()),
            ]),
            ...this.extract(this.scenarioFrom(testCaseAttempt), ({ featureDescription, rule, scenarioDescription, tags, testRunnerName }) => [
                new SceneStarts(currentSceneId, this.currentScenario, this.serenity.currentTime()),
                featureDescription && new FeatureNarrativeDetected(currentSceneId, featureDescription, this.serenity.currentTime()),
                new TestRunnerDetected(currentSceneId, testRunnerName, this.serenity.currentTime()),
                !!scenarioDescription && new SceneDescriptionDetected(currentSceneId, scenarioDescription, this.serenity.currentTime()),
                !!rule && new BusinessRuleDetected(currentSceneId, this.currentScenario, rule, this.serenity.currentTime()),
                ...tags.map(tag => new SceneTagged(currentSceneId, tag, this.serenity.currentTime())),
            ]),
        ];
    }
    parseTestStepStarted(message) {
        return this.extract(this.stepFrom(message), (step) => {
            if (this.shouldReportStep(step)) {
                const activityDetails = this.activityDetailsFor(step);
                this.currentStepActivityId = this.serenity.assignNewActivityId(activityDetails);
                return new TaskStarts(this.serenity.currentSceneId(), this.currentStepActivityId, this.activityDetailsFor(step), this.serenity.currentTime());
            }
        });
    }
    parseTestStepFinished(message) {
        return this.extract(this.stepFrom(message), (step) => {
            if (this.shouldReportStep(step)) {
                return new TaskFinished(this.serenity.currentSceneId(), this.currentStepActivityId, this.activityDetailsFor(step), this.outcomeFrom(step.result, step), this.serenity.currentTime());
            }
        });
    }
    parseTestCaseFinished(message) {
        const testCaseAttempt = this.eventDataCollector.getTestCaseAttempt(message.testCaseStartedId), currentSceneId = this.serenity.currentSceneId();
        return this.extract(this.scenarioOutcomeFrom(testCaseAttempt), ({ outcome, willBeRetried, tags }) => [
            willBeRetried ? new RetryableSceneDetected(currentSceneId, this.serenity.currentTime()) : undefined,
            ...tags.map(tag => new SceneTagged(currentSceneId, tag, this.serenity.currentTime())),
            new SceneFinished(currentSceneId, this.currentScenario, outcome, this.serenity.currentTime()),
        ]);
    }
    // ---
    extract(maybeValue, fn) {
        return (maybeValue === undefined)
            ? []
            : [].concat(fn(maybeValue)).filter(item => !!item);
    }
    scenarioDetailsFor(gherkinDocument, pickle, location) {
        return new ScenarioDetails(new Name(pickle.name), new Category(gherkinDocument.feature.name), new FileSystemLocation(this.absolutePathFrom(gherkinDocument.uri), location.line, location.column));
    }
    outlineFrom(testCaseAttempt) {
        const { gherkinDocument, pickle } = testCaseAttempt, gherkinScenarioMap = this.formatterHelpers.GherkinDocumentParser.getGherkinScenarioMap(gherkinDocument);
        if (gherkinScenarioMap[pickle.astNodeIds[0]].examples.length === 0) {
            return; // this is not an outline, skip it
        }
        const outline = gherkinScenarioMap[pickle.astNodeIds[0]];
        const details = this.scenarioDetailsFor(gherkinDocument, outline, outline.location);
        const template = new Description(outline.steps.map(step => this.testStepFormatter.format(step.keyword, step.text, step)).join('\n'));
        const examples = flatten(outline.examples.map(exampleSet => exampleSet.tableBody.map(row => ({
            header: exampleSet.tableHeader,
            row,
            name: exampleSet.name,
            description: exampleSet.description,
        })))).map((example) => ({
            rowId: example.row.id,
            name: example.name.trim(),
            description: example.description.trim(),
            values: example.header.cells
                .map(cell => cell.value)
                .reduce((values, header, i) => {
                values[header] = example.row.cells[i].value;
                return values;
            }, {}),
        }));
        const parameters = examples.find(example => example.rowId === pickle.astNodeIds.at(-1));
        return {
            details, template, parameters: new ScenarioParameters(new Name(parameters.name), new Description(parameters.description), parameters.values),
        };
    }
    scenarioFrom({ gherkinDocument, pickle }) {
        const gherkinScenarioMap = this.formatterHelpers.GherkinDocumentParser.getGherkinScenarioMap(gherkinDocument), gherkinExampleRuleMap = this.formatterHelpers.GherkinDocumentParser.getGherkinExampleRuleMap(gherkinDocument), scenarioDescription = this.formatterHelpers.PickleParser.getScenarioDescription({ gherkinScenarioMap, pickle }), scenarioTags = flatten(pickle.tags.map(tag => Tags.from(tag.name))), rule = gherkinExampleRuleMap[pickle.astNodeIds[0]];
        return {
            featureDescription: gherkinDocument.feature.description && new Description(gherkinDocument.feature.description),
            scenarioDescription: scenarioDescription && new Description(scenarioDescription),
            rule: rule && new BusinessRule(new Name(rule.name), new Description(rule.description.trim())),
            testRunnerName: new Name('JS'),
            tags: this.requirementsHierarchy.requirementTagsFor(Path.from(this.cwd).resolve(Path.from(gherkinDocument.uri)), gherkinDocument.feature.name).concat(scenarioTags),
        };
    }
    stepFrom(message) {
        const { testCaseStartedId, testStepId } = message;
        const testCaseAttempt = this.eventDataCollector.getTestCaseAttempt(testCaseStartedId);
        const index = testCaseAttempt.testCase.testSteps.findIndex(step => step.id === testStepId);
        return this.parseTestCaseAttempt(testCaseAttempt).testSteps[index];
    }
    parseTestCaseAttempt(testCaseAttempt) {
        // workaround for a bug in Cucumber 7, that's fixed in Cucumber 8 by https://github.com/cucumber/cucumber-js/pull/1531
        testCaseAttempt.testCase.testSteps.forEach(step => {
            if (!testCaseAttempt.stepResults[step.id]) {
                testCaseAttempt.stepResults[step.id] = { duration: { seconds: 0, nanos: 0 }, status: TestStepResultStatus.UNKNOWN, willBeRetried: false };
            }
        });
        // ---
        return this.formatterHelpers.parseTestCaseAttempt({
            cwd: this.cwd,
            testCaseAttempt,
            snippetBuilder: this.snippetBuilder,
            supportCodeLibrary: this.supportCodeLibrary,
        });
    }
    activityDetailsFor(parsedTestStep) {
        const location = parsedTestStep.sourceLocation || parsedTestStep.actionLocation;
        return new ActivityDetails(new Name(this.testStepFormatter.format(parsedTestStep.keyword, parsedTestStep.text || parsedTestStep.name, parsedTestStep.argument)), new FileSystemLocation(this.absolutePathFrom(location.uri), location.line));
    }
    outcomeFrom(worstResult, ...steps) {
        const Status = TestStepResultStatus;
        // todo: how does it treat failed but retryable scenarios?
        switch (worstResult.status) {
            case Status.SKIPPED:
                return new ExecutionSkipped();
            case Status.UNDEFINED: {
                const snippets = steps
                    .filter(step => step.result.status === Status.UNDEFINED)
                    .map(step => step.snippet);
                const message = snippets.length > 0
                    ? ['Step implementation missing:', ...snippets].join('\n\n')
                    : 'Step implementation missing';
                return new ImplementationPending(new ImplementationPendingError(message));
            }
            case Status.PENDING:
                return new ImplementationPending(new ImplementationPendingError('Step implementation pending'));
            case Status.AMBIGUOUS:
            case Status.FAILED: {
                const error = ErrorSerialiser.deserialiseFromStackTrace(worstResult.message);
                if (error instanceof AssertionError) {
                    return new ExecutionFailedWithAssertionError(error);
                }
                if (error instanceof TestCompromisedError) {
                    return new ExecutionCompromised(error);
                }
                return new ExecutionFailedWithError(error);
            }
            case Status.UNKNOWN:
            // ignore
            case Status.PASSED:
                return new ExecutionSuccessful();
        }
    }
    scenarioOutcomeFrom(testCaseAttempt) {
        const parsed = this.formatterHelpers.parseTestCaseAttempt({
            cwd: this.cwd,
            snippetBuilder: this.snippetBuilder,
            supportCodeLibrary: this.supportCodeLibrary,
            testCaseAttempt
        });
        const worstStepResult = parsed.testCase.worstTestStepResult;
        const willBeRetried = worstStepResult.willBeRetried || // Cucumber 7
            testCaseAttempt.willBeRetried; // Cucumber 8
        const outcome = this.outcomeFrom(worstStepResult, ...parsed.testSteps);
        const tags = [];
        if (testCaseAttempt.attempt > 0 || willBeRetried) {
            tags.push(new ArbitraryTag('retried'));
        }
        if (testCaseAttempt.attempt > 0) {
            tags.push(new ExecutionRetriedTag(testCaseAttempt.attempt));
        }
        return { outcome, willBeRetried, tags };
    }
    absolutePathFrom(relativePath) {
        return Path.from(this.cwd).resolve(Path.from(relativePath));
    }
}
function flatten(listOfLists) {
    return listOfLists.reduce((acc, current) => acc.concat(current), []);
}
