// @serenity-js/cucumber/esm/listeners/messages/parser/TestStepFormatter.js
import { createRequire as _createRequire } from "module";
const __require = _createRequire(import.meta.url);
const Table = __require("cli-table3");
/**
 * @package
 */
export class TestStepFormatter {
    format(keyword, text = '', argument) {
        return [
            // 'Given ', 'When ' and 'Then ' steps end with a space in all the supported languages - https://github.com/cucumber/gherkin/blob/fb263beaa86dd371f073277b0cc42604a82c510b/gherkin-languages.json
            // 'Before' does not have a trailing space
            keyword.endsWith(' ') || text === ''
                ? keyword
                : `${keyword}: `,
            text,
            argument && this.formatStepArgument(argument),
        ].join('').trim();
    }
    formatStepArgument(argument) {
        if (argument.docString) {
            return this.formatDocString(argument.docString);
        }
        if (argument.dataTable) {
            return this.formatDataTable(argument.dataTable);
        }
        return '';
    }
    formatDocString(docString) {
        return `\n${docString.content}`;
    }
    formatDataTable(dataTable) {
        const table = new Table({
            chars: {
                bottom: '',
                'bottom-left': '',
                'bottom-mid': '',
                'bottom-right': '',
                left: '|',
                'left-mid': '',
                mid: '',
                'mid-mid': '',
                middle: '|',
                right: '|',
                'right-mid': '',
                top: '',
                'top-left': '',
                'top-mid': '',
                'top-right': '',
            },
            style: {
                border: [],
                'padding-left': 1,
                'padding-right': 1,
            },
        });
        const rows = dataTable.rows.map(row => row.cells.map(cell => cell.value.replaceAll('\\\\', '\\\\\\\\').replaceAll('\\n', '\\\\n')));
        table.push(...rows);
        return `\n${table.toString()}`;
    }
}
