// @serenity-js/core/esm/stage/queues/DomainEventQueues.js
import { LogicError } from '../../errors/index.js';
import { SceneSequenceDetected, SceneStarts } from '../../events/index.js';
import { d } from '../../io/format.js';
import { DomainEventQueue } from './DomainEventQueue.js';
/**
 * @group Stage
 */
export class DomainEventQueues {
    queueIndex = [];
    queues = new Map();
    holdingBay = new DomainEventQueue();
    enqueue(event) {
        if (this.shouldStartNewQueueFor(event)) {
            this.queues.set(this.queueIdFor(event), new DomainEventQueue(event, ...this.holdingBay.drain()));
        }
        else if (this.hasNoQueuesReadyFor(event)) {
            this.holdingBay.enqueue(event);
        }
        else {
            this.queues.get(this.queueIdFor(event)).enqueue(event);
        }
    }
    forEach(callback) {
        this.queues.forEach(callback);
    }
    drainQueueFor(sceneId) {
        for (const [key, queue] of this.queues) {
            if (sceneId.equals(queue.sceneId)) {
                const events = queue.drain();
                this.queues.delete(key);
                return events;
            }
        }
        throw new LogicError(d `No domain event queue found for scene ${sceneId}`);
    }
    shouldStartNewQueueFor(event) {
        return (event instanceof SceneSequenceDetected || event instanceof SceneStarts)
            && !this.queues.has(this.queueIdFor(event));
    }
    hasNoQueuesReadyFor(event) {
        return this.queues.size === 0
            || !this.queues.has(this.queueIdFor(event));
    }
    queueIdFor(event) {
        const exactMatch = this.queueIndex.find(entry => entry.sceneId.equals(event.sceneId));
        if (exactMatch) {
            return exactMatch.queueId;
        }
        if (!(event instanceof SceneStarts)) {
            const sameScenarioMatch = this.queueIndex.find(entry => entry.details &&
                entry.details.equals(event.details));
            if (sameScenarioMatch) {
                this.queueIndex.push({
                    sceneId: event.sceneId,
                    details: event.details || sameScenarioMatch.details,
                    queueId: sameScenarioMatch.queueId,
                });
                return sameScenarioMatch.queueId;
            }
        }
        const newQueueId = Symbol();
        this.queueIndex.push({
            sceneId: event.sceneId,
            details: event.details,
            queueId: newQueueId,
        });
        return newQueueId;
    }
}
